### DISCUSSION TITLE
Changed Settings Tracking

### DISCUSSION THREAD

**Discussion by Fettkeewl** [arendst/Tasmota#17952] [Ideas]
Hello! 

I'm finding it ever so hard to remember all the specific settings I've fiddled with in my Tasmota Device, mainly diffrent SetOptions for various devices. 
Id like to suggest that a **Device Changed Settings** Tracking be implemented in a very basic way.

I guess it could fit in an array or a string but something that tracks what setting was altered away from the DEFAULT setting.
Example

New Tasmota Install 
* Tracking variable is empty
* User changes SO19 and SO73 away from defaults
* Tracking variable now contains the info that these settings have been altered
* User can poll the change list by a command, for instance: **ChangedSettings**
  Then gets a printout in console
  "SO19,SO73"

So only track what is changed from default, no need to track what it was changed to (to save memory ofc).
The users can simply look up the changes themselves.

**Comment by barbudor** [comment]
Default settings can be overridden by custom build. 
That kind of tracking would quickly use as much if not more than the settings itself

To manage your settings, use the tool 'decode-config' to dump the config in either JSON or Cmnd format and store each versions with a timestamp. Eventually use git for tracking and managing versions

I have a nightly script that downloads all my devices configuration and commit them into a private github repo

**Reply by Fettkeewl** [reply]
Cool Tip! Will look at it

**Reply by meingraham** [reply]
> I have a nightly script that downloads all my devices configuration and commit them into a private github repo

I have half of this but never thought of creating a repo to retain the history. Brilliant! I have a repo for my openHAB deployment... so easy peasy to add this.

**Reply by Jason2866** [reply]
@meingraham Nice to see you :-)

**Reply by sfromis** [reply]
I've got scripts for comparing `decode-config` (cached or not), output of `Status0` (has a lot of fields, some of it needs parsing), and the information from the discovery retained payloads, always quickly available. :smile:

**Reply by Fettkeewl** [reply]
> I have a nightly script that downloads all my devices configuration and commit them into a private github repo

Care to share that script or must I reinvent the wheel 😄

**Reply by barbudor** [reply]
There's nothing fancy there
You must know the IP addresses of the devices because decode-config.py talks http with the device, not mqtt

```bash
#! /usr/bin/env bash

# get latest version of decode-config
cd ~/bin
curl -s --insecure "https://raw.githubusercontent.com/tasmota/decode-config/development/decode-config.py" -o decode-config.py

# Download confs
cd ~/tasmota-conf-backups
decode-config.py -c ./my.conf -s 192.168.168.105 -o @h_@v.json
decode-config.py -c ./my.conf -s 192.168.168.106 -o @h_@v.json
... etc for each device

# remove unneeded stuff from the jsons
for conf in *.json
do
  echo cleaning file $conf
  tas-config-filter.py --input $conf --output $conf
done

# commit to git
git add *.json
git commit -m "backup $(date -Is)"
git push
```

The attached python script removes from the JSON values that are not needed like versions of decode-config etc...
That avoid to commit changes that are useless
 
[tas-config-filter.zip](https://github.com/arendst/Tasmota/files/10759699/tas-config-filter.zip)

**Reply by meingraham** [reply]
Avoids having to code all the IP addresses...

```bash
#!/bin/bash
clear

TOOLSLOC=/home/openhabian/tasmota_tools
DATALOC=$TOOLSLOC/backups/all-dmp
mkdir -p $DATALOC

for ADDR in 192.168.1.{3..254};
do
    echo $ADDR
    # Copy the app to $TOOLSLOC

    # Create dmp file and JSON file
    # If the access fails, both backup files are not created
    python3 $TOOLSLOC/decode-config/decode-config.py  -s $ADDR -o $DATALOC/config-@H-@d-@f-@v.dmp -o $DATALOC/config-@H-@d-@f-@v --json-indent 4
done
```
